### ISSUE TITLE
Coroutines (async/await/promise) language support

### ISSUE THREAD

**Issue by devlo** [zig-lang/zig#727]
Hi,

Does Zig will have language support  for coroutines or similar construct ?  Rust is going to support async/await, C++20 will get language support for coroutines and other mainstream languages (like C#, JS etc) already have/will have support too. There is already support for coroutines in LLVM  also (https://llvm.org/docs/Coroutines.html).

Any thoughts?

**Comment by andrewrk** [comment]
See also #174. I don't think it's a duplicate, because supporting coroutines is only one component of a greater concurrency story. And I do think we should support coroutines. My plan is to add experimental coroutines support as soon as 0.2.0 is released (1 week after LLVM 6.0.0 is released). That way I can work against llvm 6 instead of 5 which I believe has iterated on the coroutines support a bit.

The first implementation of coroutines in zig will likely be very low level, little more than exposing the LLVM coroutine primitives. Once myself and others have a chance to play with these tools, I'll do another design iteration on it.

**Comment by andrewrk** [comment]
As a first step towards coroutines, I submitted a patch to LLVM: https://reviews.llvm.org/D43368

In the meantime we can create an API that assumes we have this patch, and generates less efficient code (allocating too much memory) until this patch is merged and available.

**Comment by andrewrk** [comment]
cc @bnoordhuis 

I know you did (do?) a lot of libuv work, so I'd be curious to get your take on this.

**Comment by andrewrk** [comment]
Proposal edit: Renaming these builtins to keywords:
 * `@destroy()` -> `cancel` keyword
 * `@suspend()` -> `suspend` keyword
 * `@resume()` -> `resume` keyword

**Comment by andrewrk** [comment]
Proposal edit:

 * `suspend` is a block, which is given a promise as a variable in scope,
   which is the handle of the suspended async function.
 * The async function goes into the suspended state upon entry to the block,
   but it has the duration of the block to do something with the promise.
   When the async function is resumed, it will resume after the suspend block.
 * Remove `@promise()`. A `suspend` block is the only way to get access to one's own promise.

```zig
suspend |the_promise| {
    // do something with the_promise that will cause it to
    // be resumed appropriately
}
```

This corresponds to the [distinct save and suspend](http://llvm.org/docs/Coroutines.html#distinct-save-and-suspend) points in the LLVM docs.

So `waitFd` becomes:

```zig
    async fn waitFd(self: &EventLoop, fd: i32) !void {
        // A suspend block should always contain logic that will cause the
        // current async function to get resumed somehow. In this case,
        // the matching `resume` is in the `run` loop, after the `epoll_wait`.
        // We put the promise handle into epoll and this suspend point will get
        // resumed when epoll_wait says that this fd has data ready.
        suspend |the_promise| {
            var ev: epoll_event = undefined;
            ev.events = EPOLLIN|EPOLLET;
            ev.data.ptr = @ptrToInt(the_promise);
            epoll_ctl(self.epollfd, EPOLL_CTL_ADD, fd, &ev);
        }
        epoll_ctl(self.epollfd, EPOLL_CTL_DEL, fd, &ev);
    }
```

What this allows us to do is multiplex coroutines onto threads:

```zig
suspend |prom| {
    // Consider if N threads are waiting on this job queue.
    // They would call resume right away, so we must be already
    // suspended when we add the promise to the queue. This is
    // exactly what the suspend block provides.
    atomic_push_to_job_queue(prom);
}
```

**Comment by andrewrk** [comment]
Proposal edit:

 * When declaring an `async` function, if there are no parentheses after `async` then the function is generic. The allocator passed in via `async()` is used to instantiate the generic function.
 * If you use `async(AllocatorType) fn() void` syntax, then the function is not generic, and the allocator passed in will be casted to `AllocatorType`.

**Comment by bnoordhuis** [comment]
Thanks for the ping, Andrew.  Yep, I still do a lot of libuv work.

I'm intrigued by the comment on promises being linear types.  Is that a new concept in zig?

Does linearity mean the promise can be used only inside the function that creates it?  Or do you plan to introduce implicit or explicit lifetimes so that it can be returned to the caller?

Will  `@sizeOf(amain)` return the storage size of the coro?

Is `try await try async self.waitFd(fd)` equivalent to this?
```zig
var promise = try async self.waitFd(fd);
try await promise;
```
Is `try self.waitFd(fd)` without `async` a syntax error?

Will you be able to use `await afoo()` anywhere where just `foo()` is allowed?  E.g.:
```zig
if (foo()) bar();
if (await afoo()) bar();
```
Or:
```zig
for (foo()) |x| bar(x);
for (await afoo()) |x| bar(x);
for await (foo()) |x| bar(x);  // async iterator - probably a project of itself
```
I think the answer is yes, but I thought I'd ask anyway.

Does the promise start out suspended or until the first `await` or `suspend` statement?  That is relevant for CPU-bound code that you want to offload to a thread.  E.g.:
```zig
async fn everyNthPiDigit(nth: int) int { ... }
var promise = async everyNthPiDigit(1e9);  // returns immediately or after computing the billionth digit?
runInNewThread(promise);
```

I'm still a little unclear on how generators (multiple returns) would work with this proposal.

All in all, I like where this is going!

**Comment by andrewrk** [comment]
I'll answer these questions for the proposal as it stands now, but these answers are not necessarily set in stone.

> I'm intrigued by the comment on promises being linear types. Is that a new concept in zig?

Yes it is, and `promise` would be the only type that the concept applies to.

> Does linearity mean the promise can be used only inside the function that creates it? Or do you plan to introduce implicit or explicit lifetimes so that it can be returned to the caller?

The linearity is a requirement of the language but would not be entirely enforced at compile-time. You would be able to assign the promise to another variable, return it from a function, pass it as a parameter, any of these things. Each of these things would assign `false` to the `alive` property of the old reference. For example:

```zig
var promise1 = try async foo();
assert(promise1.alive);
var copy = promise1;
assert(copy.alive);
assert(!promise1.alive);
```

Under the hood, a promise is a pointer. The alive property is actually implemented by setting it to 0 to make it not alive.

Calling `cancel`, `resume`, or `await` on a promise which is not alive is undefined behavior - that is - in Debug and ReleaseSafe builds it panics, and in ReleaseFast mode the optimizer assumes you never do it to make whatever optimizations it can.

So it's a linear type in the sense of what the responsibility is for the programmer - one must `cancel`, `resume`, or `await` a promise exactly once, otherwise undefined behavior occurs.

In practice, if one fails to do anything with a promise, then the async function's defers, if any, will not run. I haven't thought of a way to provide any runtime safety checks for this situation. It's pretty similar to not calling free, which is difficult to catch in the same way.

It's also possible to cause a promise to have an incorrect value for `alive`, by using unsafe features such as `@memcpy` instead of `std.mem.copy`:

 * store an alive `promise` in a struct
 * `@ptrCast` to get a `&u8`
 * `@memcpy` to copy the struct to another one
 * both structs now have a promise that thinks it's alive

Because of this, I would consider actually removing the ability to query the `alive` state from the proposal. Programmers would have to manage such state themselves if they wanted it, but we would still insert runtime safety checks, which would have the possibility of false negatives, but no false positives. (And then in ReleaseFast we would not even bother setting the pointer to 0 since it can never be queried.)

Either way, the caller can, in fact, return the promise.

> Will `@sizeOf(amain)` return the storage size of the coro?

That's related to the code review I sent to LLVM above. I actually abandoned it, because the proposal I ended up coming up with didn't necessarily need it. However, now that you bring it up again, it would be nice to have the ability to query the coroutine frame size. Unfortunately, without significant reorganization of how zig works, and tightly coupling zig to LLVM instead of loose coupling as it currently stands, the size would not be available to use at `comptime`.

Also, currently when you reference a function, it's actually a comptime known pointer to a function. So `@sizeOf(amain)` would be `@sizeOf(usize)`. We'd have to introduce a new builtin such as `@frameSize`. This is related to #157.

> Is `try await try async self.waitFd(fd)` equivalent to this?

Yes.

> Is `try self.waitFd(fd)` without `async` a syntax error?

It's a semantic analysis error, and in fact is already implemented in the `async` branch: 

```zig
test "aoeu" {
    amain();
}
async fn amain() void {
}
```

Output:

```
[nix-shell:~/dev/zig/build]$ ./zig test test2.zig 
/home/andy/dev/zig/build/test2.zig:2:5: error: must use async keyword to call async function
    amain();
    ^
/home/andy/dev/zig/build/test2.zig:5:7: note: declared here
async fn amain() void {
      ^
```

> Will you be able to use `await afoo()` anywhere where just `foo()` is allowed?  E.g.:

You would get `error: expected promise, found 'i32'` or similar.

> Does the promise start out suspended or until the first `await` or `suspend` statement?  That is relevant for CPU-bound code that you want to offload to a thread.  E.g.:

No - with this proposal the async function executes until suspension. You would have to intentionally offload CPU-bound code to a thread like this:

```zig
async fn everyNthPiDigit(nth: int) int {
    suspend |my_promise| {
        runInNewThread(my_promise);
    }
    // we are now running in the other thread
    // do computation...
    suspend |my_promise| {
        runInOriginalThread(my_promise);
    }
    // back in the original thread
    return result;
}
```

Some options here:
 * You could let the other thread complete the async function, if your application can handle multiple threads. That would be more efficient.
 * You could have the calling function do the `suspend` dance and then call `everyNthPiDigit` which could be thread agnostic. Using comptime features you could also make a function with var args to call another async function in another thread, such as `async fn callAsyncFnInNewThread(func: var, args: ...) @typeOf(func).ReturnType`
 * (my favorite) Instead of having async functions have to care about kernel threads, you could have the main event loop use a thread pool to execute waiting promises. For example the function in the proposal that calls `epoll_wait` could use a thread pool to resume N promises at once.
    
    In a busy web server, for example, let's say 5 requests came in to calculate Pi on a 4 processor core machine. Thread pool size 4. The first request comes in, thread 1 picks it up and it gets stuck in `everyNthPiDigit`. Then `epoll_wait` reveals another request comes in, so even though thread 1 is stuck in `everyNthPiDigit`, thread 2 picks up the waiting fd, and goes for it, then gets stuck in `everyNthPiDigit`. So on until all 4 threads are stuck in `everyNthPiDigit`. Now the 5th request comes in, and there are no threads ready to pick up the ready fds. All 4 threads are stuck in the CPU bound `everyNthPiDigit`. You could change the web server to have something other than this behavior, but this is maximally efficient. To context switch away from this CPU bound work would be to sacrifice performance.

The eager execution of async functions does the following things:
 * saves 2 context switches per call
 * gives the async function a chance to start its own promises as soon as possible (what if the calling function has CPU-bound code in between `async` and `await` of the async fn?)
 * avoids the necessity of the async/await language features having to be integrated with a userland event loop code

> I'm still a little unclear on how generators (multiple returns) would work with this proposal.

To be determined. As is, this proposal does not support multiple returns.

**Comment by andrewrk** [comment]
Here's an example of what `amain` would look like if we did not give the `alive` property of a promise:

```zig
async fn amain(loop: &EventLoop) !void {
    const config_promise = try async loop.allocReadFile("config.json");
    var config_promise_alive = true;
    defer if (config_promise_alive) cancel config_promise;

    const dict_promise = try async loop.allocReadFile("dictionary.txt");
    var dict_promise_alive = true;
    defer if (dict_promise_alive) cancel dict_promise;

    config_promise_alive = false;
    const config_contents = try await config_promise;
    defer allocator.free(config_contents);

    dict_promise_alive = false;
    const dict_contents = try await dict_promise;
    defer allocator.free(dict_contents);

    if (mem.eql(u8, dict_contents, config_contents)) {
        std.debug.warn("config.json and dictionary.txt are the same");
    }
}
```

It's a bit verbose, but it does not introduce nesting, it's pretty straightforward, and does not introduce any new language features. My biggest concern would be that it would be easier to not do this than to do it, thus leading to incorrect cleanup code.

I have a more general dont-forget-to-cleanup-resources proposal coming up soon that could be integrated into this.

**Comment by andrewrk** [comment]
With #782 the example above would become:

```zig
async fn amain(loop: &EventLoop) !void {
    const config_promise = clean try async loop.allocReadFile("config.json");
    const dict_promise = clean try async loop.allocReadFile("dictionary.txt");

    const config_contents = clean try await config_promise;
    const dict_contents = clean try await dict_promise;

    if (mem.eql(u8, dict_contents, config_contents)) {
        std.debug.warn("config.json and dictionary.txt are the same");
    }
}
```

**Comment by 0joshuaolson1** [comment]
In case this is helpful to someone: [some C++ async/await/resumable proposals](https://gist.github.com/jamboree/a2b3fe32eeb8c21e820c)

**Comment by Meai** [comment]
I dont see how async/await fits into zig's philosophy of not hiding function calls, there is now an entire threadpool (right?!) hidden inside a keyword with complex stack rebuilding and what not. I cant tell you how to be threadsafe and write performant code while there are countless threads doing God knows what;

**Comment by andrewrk** [comment]
> there is now an entire threadpool

There's no thread pool in the zig language. Coroutines work in freestanding mode, with no OS.

Here's the thread pool: https://github.com/ziglang/zig/blob/fd575fe1f3b45806f2cf823a2abe3727d381d4ed/std/event/loop.zig#L16

It's in userland. If you don't import `std.event.Loop`, then you don't get a thread pool.
